firmware/library/L1_Drivers/can.hpp

```diff
@@ -1,49 +1,986 @@
+        can_registers[controller_]->IER |= (1 << kTxBuffer3IntEnableBit);
+    }
+
+    void SetBaudRate()
+    {
+        // CAN Baud Rate : Assume 100kbps (100Khz)
+        //
+        // According to the BOSCH CAN spec, the nominal bit time
+        // is divided into 4 time segments. These segments need to
+        // be programmed for the internal bit timing logic/state machine.
+        // Refer to the link below for more details about the importance
+        // of these time segments:
+        // http://www.keil.com/dd/docs/datashts/silabs/boschcan_ug.pdf
+        //
+        // Nominal Bit Time : 1 / 100 000 == 10^-5s
+        //
+        //   0_______________________________10^-5
+        //  _/             1 bit             \_
+        //   \_______________________________/
+        //
+        //   | SYNC | PROP | PHASE1 | PHASE2 |        BOSCH
+        //   | TSCL |     TSEG1     | TSEG2  |        LPC
+        //                          ^
+        //                    sample point (From Vector: 80%)
+        //
+        // TODO(rabeelio): Add detailed explanation for below:
+        // The sync seg
+        // THe prop seg
+        // phase1 seg
+        // phase 2 seg
+        //
+        // Each segment is a # of time quanta (Tq). Per spec: (8 <= Tq <= 25)
+        // A Tq is derived from the CAN System Clock and Baud Rate
+        //
+        // | SJW | (Re-)Synchronization Jump Width
+        //
+        // The SJW is used to compensate for positive and negative
+        // edge phase errors; it defines the max Tq that the
+        // bit timing logic can shift the sample point.
+        //
+        //
+        // Compute Tq:
+        //
+        // APB Clock (Peripheral Clock) : Assume 48Mhz
+        //
+        // _||_||_||_||_||_||_
+        //
+        // Apply Baud Rate Prescaler (BRP) | Assume (BRP) = 48
+        //
+        // Result: CAN System Clock | Result: 1Mhz
+        //   _    _    _    _
+        // _| |__| |__| |__| |_ :
+        //
+        // Tq = CAN System Clock / Baud Rate
+        // Tq = 1 000 000 / 100 000 = 10
+        //
+        // The following labels are how LPC interprets
+        // the CAN spec as detailed above.
+        //
+        // LPC:      BOSCH:
+        // SJW   =   SJW           (PROGRAMMABLE)
+        // TSCL  =   SYNC          (FIXED TO 1TQ)
+        // TSEG1 =   PROP + PHASE1 (PROGRAMMABLE)
+        // TSEG2 =   PHASE2        (PROGRAMMABLE)
+        //
+        // | TSCL |     TSEG1     | TSEG2  |
+        // | 1Tq  |      7Tq      |  2Tq   |
+        //
+        // | SJW  |
+        // | 3Tq  |
+
+        // 10Tq = 1Tq + 7Tq + 2Tq
+        // Nominal Bit Time = TSCL + TSEG1 + TSEG2
+        // 1/100kbit/s = 10^-5s = (1Tq * 10^-6) +
+        //                        (7Tq * 10^-6) +
+        //                        (2Tq * 10^-6)
+        if (baud_rate_ == kBaud100Kbps)
+        {
+            can_registers[controller_]->BTR = (0 << 23) | (1 << 20) |
```

Comment:
This would be a great place for a bit field.

Reply:
done